### PR TITLE
Add Depth Anything V2 Metric models

### LINKED ISSUE #31972: Add Depth Anything v2 metric depth
### Feature request

We already have [Depth Anything v2](https://huggingface.co/docs/transformers/main/en/model_doc/depth_anything_v2) available in Transformers (compatible with v1). These are relative depth estimation models, capable of predicting the relative distance among pixels.

There are also absolute (also called metric) depth estimation models available here: https://github.com/DepthAnything/Depth-Anything-V2/tree/main/metric_depth. Would be great to make these also Transformers compatible.

The author also informed me that they are only minor differences compared to their relative counterparts:

> Our model definition for metric depth model in V2 is almost similar to our relative depth model: https://github.com/LiheYoung/Depth-Anything/blob/main/depth_anything/dpt.py. The only difference is we replace the final ReLU(out) (https://github.com/LiheYoung/Depth-Anything/blob/1e1c8d373ae6383ef6490a5c2eb5ef29fd085993/depth_anything/dpt.py#L99) to a sigmoid(out) * max_depth.

### Motivation

We currently only have one metric depth estimation model available, namely [ZoeDepth](https://huggingface.co/docs/transformers/main/en/model_doc/zoedepth). Would be great to expand this to more SOTA models

### Your contribution

Could work on this, but it's a great opportunity for a first open-source contribution!

### PR DESCRIPTION
**Pull request by bt2513**
# What does this PR do?

Add Depth Anything V2 models for _metric_ depth estimations

Fixes #31972 

## Before submitting
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@NielsRogge @amyeroberts

### CODE REVIEW

**Comment by bt2513** [comment]
Hi @NielsRogge, I've just opened a PR to add the models. I plan to actively work on it this week. As this is my first contribution, I'll rely on your guidance if needed. Thank you! :)

**Review by Khaliq88 — CHANGES_REQUESTED** [review]
Traceback (most recent call last):
  File "/home/jon/miniconda3/envs/h2ogpt/lib/python3.10/site-packages/urllib3/connection.py", line 198, in _new_conn
    sock = connection.create_connection(
  File "/home/jon/miniconda3/envs/h2ogpt/lib/python3.10/site-packages/urllib3/util/connection.py", line 60, in create_connection
    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
  File "/home/jon/miniconda3/envs/h2ogpt/lib/python3.10/socket.py", line 955, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -3] Temporary failure in name resolution

**Comment by bt2513** [comment]
Hi @NielsRogge - I've adapted the `DepthAnythingConfig` to support _metric_ depth estimation by introducing an attribute that enables easy switching of the model configuration. I've also modified the `convert_depth_anything_to_hf.py` script to validate my results against the original repo, and things look good.

As this is my first time contributing to add a model, I'd greatly appreciate your guidance on what steps to take next to get this merged. Thank you!

File: src/transformers/models/depth_anything/configuration_depth_anything.py
Comment (NielsRogge) [inline src/transformers/models/depth_anything/configuration_depth_anything.py:71]: Perhaps we can rename this to either
* `depth_estimation_type`, to indicate that it's a type
* or `final_activation_function``? Which can be either "relu" or "sigmoid"
cc @amyeroberts

**Review by NielsRogge — COMMENTED** [review]
PR is looking very nice! Did you confirm all slow tests are passing (RUN_SLOW=yes pytest tests/models/depth_anything)?

In that case, we can also add a slow integration test for the metric depth estimation checkpoints, similar to [here](https://github.com/huggingface/transformers/blob/49928892d6491ff5a49c12cbc34695f6fa7ac0ed/tests/models/depth_anything/test_modeling_depth_anything.py#L248).

File: src/transformers/models/depth_anything/configuration_depth_anything.py
Comment (bt2513) [reply src/transformers/models/depth_anything/configuration_depth_anything.py:71]: Thanks for the feedback @NielsRogge! 

I think it makes more sense to use `depth_estimation_type` than `final_activation_function` since using a "metric" depth estimation type encompasses more than just switching the final activation function by Sigmoid - it also involves scaling the output by `max_depth`. So, I believe using `depth_estimation_type` would be more accurate for this concept

**Comment by bt2513** [comment]
> PR is looking very nice! Did you confirm all slow tests are passing (RUN_SLOW=yes pytest tests/models/depth_anything)?
> 
> In that case, we can also add a slow integration test for the metric depth estimation checkpoints, similar to [here](https://github.com/huggingface/transformers/blob/49928892d6491ff5a49c12cbc34695f6fa7ac0ed/tests/models/depth_anything/test_modeling_depth_anything.py#L248).

Good catch, I actually forgot to run these tests! And they are not passing ... I don't think the failures are related to my changes though, as they seem about the backbone initialization:

Screenshot 2024-07-29 at 6 35 19 PM

I'll investigate this and try to fix the failure

**Comment by bt2513** [comment]
If I am reproducing the error correctly, which I replicated with the following code by retracing what happens in the code I believe:
```
from transformers import TimmBackbone, Dinov2Config, AutoBackbone, TimmBackboneConfig
import timm

image_size=32
patch_size=16
hidden_size=4
num_hidden_layers=2
num_attention_heads=2
intermediate_size=8
is_training=True
out_features=["stage1", "stage2"]
reshape_hidden_states=False
num_channels=3

config = Dinov2Config(
    image_size=image_size,
    patch_size=patch_size,
    num_channels=num_channels,
    hidden_size=hidden_size,
    num_hidden_layers=num_hidden_layers,
    num_attention_heads=num_attention_heads,
    intermediate_size=intermediate_size,
    is_training=is_training,
    out_features=out_features,
    reshape_hidden_states=reshape_hidden_states,
)

config.backbone = "resnet18"
config.use_pretrained_backbone = True
config.use_timm_backbone = True
config.backbone_config = None
# For transformer backbones we can't set the out_indices or just return the features
config.backbone_kwargs = {"out_indices": (-2, -1)}

# backbone = AutoBackbone.from_pretrained(
#         config.backbone,
#         use_timm_backbone=config.use_timm_backbone,
#         use_pretrained_backbone=config.use_pretrained_backbone,
#         **config.backbone_kwargs,
#     )

backbone_config = getattr(config, "backbone_config", None)
use_timm_backbone = getattr(config, "use_timm_backbone", None)
use_pretrained_backbone = getattr(config, "use_pretrained_backbone", None)
backbone_checkpoint = getattr(config, "backbone", None)
backbone_kwargs = getattr(config, "backbone_kwargs", None)
backbone_kwargs = {} if backbone_kwargs is None else backbone_kwargs

backbone_config = TimmBackboneConfig()
num_channels = backbone_config.num_channels
features_only = backbone_config.features_only
use_pretrained_backbone = backbone_config.use_pretrained_backbone
out_indices = backbone_config.out_indices
config = TimmBackboneConfig(
    backbone=config.backbone,
    num_channels=num_channels,
    features_only=features_only,
    use_pretrained_backbone=use_pretrained_backbone,
    out_indices=out_indices,
)

# TimmBackbone(config)

out_indices = config.out_indices if getattr(config, "out_indices", None) is not None else (-1,)
_backbone = timm.create_model(
        config.backbone,
        pretrained=config.use_pretrained_backbone,
        # This is currently not possible for transformer architectures.
        features_only=config.features_only,
        in_chans=config.num_channels,
        out_indices=out_indices,
    )

# _return_layers = {
#         layer["module"]: str(layer["index"]) for layer in _backbone.feature_info.get_dicts()
#     }
print(_backbone.feature_info.get_dicts())
```
The error is due because `_backbone.feature_info.get_dicts()` seems to return `[{'num_chs': 512, 'reduction': 32, 'module': 'layer4'}]`, and there is indeed no key "index" in the dict. Any guidance on how to fix this issue? I am not sure what this attribute should return ideally in [modeling_timm_backbone.py](https://github.com/huggingface/transformers/blob/f0bc49e7f61f74f055c47ad40e6010f57eed0b0b/src/transformers/models/timm_backbone/modeling_timm_backbone.py#L85)

**Comment by bt2513** [comment]
If I comment out the timm backbone test (i.e. [these lines](https://github.com/huggingface/transformers/blob/main/tests/models/depth_anything/test_modeling_depth_anything.py#L220-L227)), the tests passed successfully otherwise :)

File: src/transformers/models/chameleon/image_processing_chameleon.py
Comment (amyeroberts) [inline src/transformers/models/chameleon/image_processing_chameleon.py:83]: nit - convention is to not have numeric values in code formatting for the docstrings 

```suggestion
        resample (`PILImageResampling`, *optional*, defaults to 1):
```

File: src/transformers/models/depth_anything/configuration_depth_anything.py
Comment (amyeroberts) [inline src/transformers/models/depth_anything/configuration_depth_anything.py:71]: We should validate this within the init

File: src/transformers/models/depth_anything/configuration_depth_anything.py
Comment (amyeroberts) [inline src/transformers/models/depth_anything/configuration_depth_anything.py:150]: Could we use a smarter default here? Given 20 is for indoor models, that might make more sense.

File: src/transformers/models/depth_anything/modeling_depth_anything.py
Comment (amyeroberts) [inline src/transformers/models/depth_anything/modeling_depth_anything.py:340]: Nice :)

File: tests/models/depth_anything/test_modeling_depth_anything.py
Comment (amyeroberts) [inline tests/models/depth_anything/test_modeling_depth_anything.py:269]: What's pending for the checkpoint to be made available?

**Review by amyeroberts — APPROVED** [review]
Looks great - thanks for adding this! 

Just some small nits. Once we have the checkpoints updated for the tests we can merge!

File: src/transformers/models/chameleon/image_processing_chameleon.py
Comment (bt2513) [reply src/transformers/models/chameleon/image_processing_chameleon.py:83]: I can revert this, no problem - it was actually changed after running `make fix-copies`

File: src/transformers/models/depth_anything/configuration_depth_anything.py
Comment (bt2513) [reply src/transformers/models/depth_anything/configuration_depth_anything.py:150]: It actually has to be 1 by default, if we are not using a _metric_ depth model, because in that case, the output shouldn't be scaled by `max_depth`. So, if `depth_estimation_type` is `relative`, it should be 1

File: tests/models/depth_anything/test_modeling_depth_anything.py
Comment (bt2513) [reply tests/models/depth_anything/test_modeling_depth_anything.py:269]: I just don't know how to convert the checkpoints under the organization name 😅 Could you please provide some guidance for this? When I run the [convert_depth_anything_to_hf.py](https://github.com/huggingface/transformers/blob/f0bc49e7f61f74f055c47ad40e6010f57eed0b0b/src/transformers/models/depth_anything/convert_depth_anything_to_hf.py) script, it pushes the checkpoints on the hub under my workspace

**Comment by bt2513** [comment]
> If I am reproducing the error correctly, which I replicated with the following code by retracing what happens I believe:
> 
> ```
> from transformers import TimmBackbone, Dinov2Config, AutoBackbone, TimmBackboneConfig
> import timm
> 
> image_size=32
> patch_size=16
> hidden_size=4
> num_hidden_layers=2
> num_attention_heads=2
> intermediate_size=8
> is_training=True
> out_features=["stage1", "stage2"]
> reshape_hidden_states=False
> num_channels=3
> 
> config = Dinov2Config(
>     image_size=image_size,
>     patch_size=patch_size,
>     num_channels=num_channels,
>     hidden_size=hidden_size,
>     num_hidden_layers=num_hidden_layers,
>     num_attention_heads=num_attention_heads,
>     intermediate_size=intermediate_size,
>     is_training=is_training,
>     out_features=out_features,
>     reshape_hidden_states=reshape_hidden_states,
> )
> 
> config.backbone = "resnet18"
> config.use_pretrained_backbone = True
> config.use_timm_backbone = True
> config.backbone_config = None
> # For transformer backbones we can't set the out_indices or just return the features
> config.backbone_kwargs = {"out_indices": (-2, -1)}
> 
> # backbone = AutoBackbone.from_pretrained(
> #         config.backbone,
> #         use_timm_backbone=config.use_timm_backbone,
> #         use_pretrained_backbone=config.use_pretrained_backbone,
> #         **config.backbone_kwargs,
> #     )
> 
> backbone_config = getattr(config, "backbone_config", None)
> use_timm_backbone = getattr(config, "use_timm_backbone", None)
> use_pretrained_backbone = getattr(config, "use_pretrained_backbone", None)
> backbone_checkpoint = getattr(config, "backbone", None)
> backbone_kwargs = getattr(config, "backbone_kwargs", None)
> backbone_kwargs = {} if backbone_kwargs is None else backbone_kwargs
> 
> backbone_config = TimmBackboneConfig()
> num_channels = backbone_config.num_channels
> features_only = backbone_config.features_only
> use_pretrained_backbone = backbone_config.use_pretrained_backbone
> out_indices = backbone_config.out_indices
> config = TimmBackboneConfig(
>     backbone=config.backbone,
>     num_channels=num_channels,
>     features_only=features_only,
>     use_pretrained_backbone=use_pretrained_backbone,
>     out_indices=out_indices,
> )
> 
> # TimmBackbone(config)
> 
> out_indices = config.out_indices if getattr(config, "out_indices", None) is not None else (-1,)
> _backbone = timm.create_model(
>         config.backbone,
>         pretrained=config.use_pretrained_backbone,
>         # This is currently not possible for transformer architectures.
>         features_only=config.features_only,
>         in_chans=config.num_channels,
>         out_indices=out_indices,
>     )
> 
> # _return_layers = {
> #         layer["module"]: str(layer["index"]) for layer in _backbone.feature_info.get_dicts()
> #     }
> print(_backbone.feature_info.get_dicts())
> ```
> 
> The error is due because `_backbone.feature_info.get_dicts()` seems to return `[{'num_chs': 512, 'reduction': 32, 'module': 'layer4'}]`, and there is indeed no key "index" in the dict. Any guidance on how to fix this issue? I am not sure what this attribute should return ideally in [modeling_timm_backbone.py](https://github.com/huggingface/transformers/blob/f0bc49e7f61f74f055c47ad40e6010f57eed0b0b/src/transformers/models/timm_backbone/modeling_timm_backbone.py#L85)

Thanks for the review @amyeroberts! 🙌 I'll address your comments in a bit :)
Any insights on the above? This currently prevents the slow tests for depth anything to pass, and I am not sure how to resolve it.  cc @NielsRogge

File: src/transformers/models/depth_anything/configuration_depth_anything.py
Comment (amyeroberts) [reply src/transformers/models/depth_anything/configuration_depth_anything.py:154]: Ah, yes, you're right!

File: tests/models/depth_anything/test_modeling_depth_anything.py
Comment (dvksn) [inline tests/models/depth_anything/test_modeling_depth_anything.py:256]: Hey @bt2513, first time contributor here. I was running your code locally and found that the order of args is not correct here.

File: tests/models/depth_anything/test_modeling_depth_anything.py
Comment (NielsRogge) [reply tests/models/depth_anything/test_modeling_depth_anything.py:269]: I can transfer them from your HF profile to the organization

File: tests/models/depth_anything/test_modeling_depth_anything.py
Comment (bt2513) [reply tests/models/depth_anything/test_modeling_depth_anything.py:256]: Thanks for catching this! I made some last min cleanup before pushing and missed it 😅

File: tests/models/depth_anything/test_modeling_depth_anything.py
Comment (bt2513) [reply tests/models/depth_anything/test_modeling_depth_anything.py:269]: Thanks @NielsRogge! Here's the list of checkpoints to transfer:
-  `bthia97/Depth-Anything-V2-Metric-Indoor-Small-hf`
- `bthia97/Depth-Anything-V2-Metric-Indoor-Base-hf`
- `bthia97/Depth-Anything-V2-Metric-Indoor-Large-hf`
- `bthia97/Depth-Anything-V2-Metric-Outdoor-Small-hf`
- `bthia97/Depth-Anything-V2-Metric-Outdoor-Base-hf`
- `bthia97/Depth-Anything-V2-Metric-Outdoor-Large-hf`

Am I in charge of designing the model cards? Please let me know if I need to do it! :)

**Comment by bt2513** [comment]
Hi @NielsRogge @amyeroberts, I've updated the checkpoints in the tests. Once they are transferred from my account to the organization, things should be good I believe. Please let me know if you have any additional feedback :)

**Comment by amyeroberts** [comment]
@bt2513 Great - I'll let @NielsRogge let us know when the checkpoints have been updated :) 

Is the error [with the backbone](https://github.com/huggingface/transformers/pull/32126#issuecomment-2259145967) still an issue?

**Comment by bt2513** [comment]
> @bt2513 Great - I'll let @NielsRogge let us know when the checkpoints have been updated :)
> 
> Is the error [with the backbone](https://github.com/huggingface/transformers/pull/32126#issuecomment-2259145967) still an issue?

@amyeroberts Yes, it is! It's actually not resolved yet - my bad, I had this part commented out to ensure that the other tests passed

File: tests/models/depth_anything/test_modeling_depth_anything.py
Comment (NielsRogge) [reply tests/models/depth_anything/test_modeling_depth_anything.py]: Thanks, I can take care of this tonight. Feel free to add model cards

Comment (bt2513) [reply tests/models/depth_anything/test_modeling_depth_anything.py]: Done! I took example on the cards you created for Depth Anything V2 (relative depth metric) :)

**Comment by bt2513** [comment]
> > @bt2513 Great - I'll let @NielsRogge let us know when the checkpoints have been updated :)
> > Is the error [with the backbone](https://github.com/huggingface/transformers/pull/32126#issuecomment-2259145967) still an issue?
> 
> @amyeroberts Yes, it is! It's actually not resolved yet - my bad, I had this part commented out to ensure that the other tests passed. Could you please help me with this? I am not sure what the attribute should return ideally in [modeling_timm_backbone.py](https://github.com/huggingface/transformers/blob/f0bc49e7f61f74f055c47ad40e6010f57eed0b0b/src/transformers/models/timm_backbone/modeling_timm_backbone.py#L85), so it's hard for me to debug

Actually, all the tests in [test_modeling_timm_backbone.py](https://github.com/huggingface/transformers/blob/main/tests/models/timm_backbone/test_modeling_timm_backbone.py) also fail for me when I run them. I'm not sure if this is an issue with my local setup, if I am missing something, or if something has changed
Screenshot 2024-08-02 at 9 19 56 PM
Do they pass for you @amyeroberts @NielsRogge? It's the same errors I get when running `test_modeling_depth_anything` (as mentioned in my previous comments), so if you don't encounter this issue, it might be that the tests are actually fine and this is just an issue with my setup

File: tests/models/depth_anything/test_modeling_depth_anything.py
Comment (NielsRogge) [reply tests/models/depth_anything/test_modeling_depth_anything.py]: Thanks, just transferred them all :)

**Comment by bt2513** [comment]
Thanks for transferring the weights, @NielsRogge! The only PR blocker now is the issue mentioned above. Can you help resolve this? 🙏 Once it's fixed, we should be able to merge 🥳
cc @amyeroberts

**Comment by bt2513** [comment]
> Hi @bt2513 it can be solved by running `pip install --upgrade timm`.
> 
> All tests pass for me.

Oh, thanks @NielsRogge! You're right, I thought I had done this, my bad 🤦🏻‍♂️
I believe everything looks good then :) I don't have merge permissions, so I am deferring to you for the next steps. Thanks!

File: src/transformers/models/depth_anything/configuration_depth_anything.py
Comment (amyeroberts) [inline src/transformers/models/depth_anything/configuration_depth_anything.py:152]: This should be an exception, rather than an assert

File: tests/models/depth_anything/test_modeling_depth_anything.py
Comment (amyeroberts) [inline tests/models/depth_anything/test_modeling_depth_anything.py:244]: Let's move the predict logic back into the test and keep with `prepare_img` -- this is more consistent with other tests and doesn't abstract away the important part: it's better for tests to be clear and explicit

**Comment by bt2513** [comment]
Got it @amyeroberts, that makes sense. I just made the changes to address your feedback, please let me know if there are any other things you'd like me to change :) thanks!

**Comment by bt2513** [comment]
Hi @amyeroberts @NielsRogge, bumping this in case you missed my last changes :)

**Review by NielsRogge — APPROVED** [review]
Thanks a lot for adding this, great work!

**Comment by amyeroberts** [comment]
@bt2513 Looks great - thanks for iterating! Only thing left to do is to run the slow tests on our CI runners. Could you push a commit with `[run_slow] depth_anything` as a messsage? This should trigger the tests (I'll need to approve the work flow). We may need to update the integration test values a little as there can be slight numerical differences given hardware and envs

**Comment by bt2513** [comment]
@amyeroberts it looks like the `PR slow CI / Run all tests for the model (models/depth_anything, single-gpu) (pull_request)` job failed, but I can't see any error message when I click on details - except "This job failed". Should we just try to re-run it or is there anything else you can see that can help fix it? Please let me know :)

**Comment by NielsRogge** [comment]
@bt2513 thanks, I reported it to team, the infrastructure is being updated. Will rerun when that's finished

**Comment by NielsRogge** [comment]
Congrats on this amazing contribution @bt2513 :) would be great to share about it on social media, which we'll help amplify!

**Pull request merged**
